File: src/com/zoarial/iot/threads/tcp/SocketHelper.java
package com.zoarial.iot.threads.tcp;

import com.zoarial.PrintBaseClass;
import com.zoarial.iot.models.IoTSession;
import com.zoarial.iot.models.actions.IoTAction;
import com.zoarial.iot.models.actions.IoTActionArgument;
import com.zoarial.iot.models.actions.IoTActionArgumentList;

import java.io.*;
import java.net.Inet4Address;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.util.Arrays;
import java.util.UUID;

class SocketHelper extends PrintBaseClass {
    public final Socket inSocket;
    public final BufferedOutputStream rawOut;
    public final DataOutputStream out;
    public final BufferedInputStream rawIn;
    public final DataInputStream in;
    private final boolean encrypted;
    private final boolean local;

    public SocketHelper(Socket socket) {
        super("Socket Helper");
        inSocket = socket;
        DataOutputStream tempOut = null;
        DataInputStream tempIn = null;
        BufferedInputStream tempRawIn = null;
        BufferedOutputStream tempRawOut = null;

        if(socket.getRemoteSocketAddress().toString().startsWith("/127.")) {
            encrypted = true; // Its a local connection. Same effect as encryption.
            local = true;
        } else {
            encrypted = false;
            local = false;
        }

        try {
            tempRawOut = new BufferedOutputStream(inSocket.getOutputStream());
            tempOut = new DataOutputStream(tempRawOut);
            tempRawIn = new BufferedInputStream(inSocket.getInputStream());
            tempIn = new DataInputStream(tempRawIn);
        } catch (IOException ex) {
            System.out.println("Something happened while creating inSocketWrapper. Exiting.");
            System.exit(-1);
        }
        out = tempOut;
        in = tempIn;
        rawOut = tempRawOut;
        rawIn = tempRawIn;
    }

    public String readKey() throws IOException {
        StringBuilder str = new StringBuilder();
        byte b = in.readByte();
        while(b != ':') {
            str.append((char)b);
            b = in.readByte();
        }
        return str.toString();
    }

    public String readValue() throws  IOException {
        StringBuilder str = new StringBuilder();
        byte b = in.readByte();
        while(b != ',' && b != '.') {
            str.append((char)b);
            b = in.readByte();
        }
        return str.toString();
    }

    public String readString() throws IOException {
        StringBuilder str = new StringBuilder();
        byte b = in.readByte();
        while(b != 0) {
            str.append((char)b);
            b = in.readByte();
        }
        return str.toString();
    }

    public UUID readUUID() throws IOException {
        return new UUID(in.readLong(), in.readLong());
    }

    public int readInt() throws IOException {
        return in.readInt();
    }

    public long readLong() throws IOException {
        return in.readLong();
    }

    public byte readByte() throws IOException {
        return in.readByte();
    }

    public IoTActionArgumentList readArgumentList(IoTAction action) throws IOException {

        IoTActionArgumentList argumentList = new IoTActionArgumentList();

        byte b = in.readByte();
        while(b != '.') {
            StringBuilder str = new StringBuilder();
            while (b != ',' && b != '.') {
                str.append((char) b);
                b = in.readByte();
            }
            argumentList.add(new IoTActionArgument(str.toString()));
        }
        return argumentList;
    }


    public boolean isEncrypted() {
        return encrypted;
    }

    public boolean isLocal() {
        return local;
    }

    public boolean isClosed() {
        return inSocket.isClosed();
    }

    public void close() throws IOException {
        out.close();
        in.close();
        inSocket.close();
    }

}

File: src/main/java/com/zoarial/iot/threads/tcp/SocketHelper.kt
package com.zoarial.iot.threads.tcp

import com.zoarial.PrintBaseClass
import me.zoarial.networkArbiter.ZoarialNetworkArbiter
import java.io.*
import java.net.Socket
import java.util.*
import kotlin.system.exitProcess

class SocketHelper(val inSocket: Socket) : PrintBaseClass("Socket Helper") {
    val rawOut: BufferedOutputStream
    val out: DataOutputStream
    val rawIn: BufferedInputStream
    val `in`: DataInputStream
    var isEncrypted = false
    var isLocal = false
    private var networkArbiter: ZoarialNetworkArbiter = ZoarialNetworkArbiter

    init {
        var tempOut: DataOutputStream? = null
        var tempIn: DataInputStream? = null
        var tempRawIn: BufferedInputStream? = null
        var tempRawOut: BufferedOutputStream? = null
        if (inSocket.remoteSocketAddress.toString().startsWith("/127.")) {
            isEncrypted = true // It's a local connection. Same effect as encryption.
            isLocal = true
        } else {
            isEncrypted = false
            isLocal = false
        }
        try {
            tempRawOut = BufferedOutputStream(inSocket.getOutputStream())
            tempOut = DataOutputStream(tempRawOut)
            tempRawIn = BufferedInputStream(inSocket.getInputStream())
            tempIn = DataInputStream(tempRawIn)
        } catch (ex: IOException) {
            println("Something happened while creating inSocketWrapper. Exiting.")
            exitProcess(-1)
        }
        out = tempOut
        `in` = tempIn
        rawOut = tempRawOut
        rawIn = tempRawIn
    }

    @Throws(IOException::class)
    fun readKey(): String {
        val str = StringBuilder()
        var b = `in`.readByte()
        while (b != ':'.code.toByte()) {
            str.append(Char(b.toUShort()))
            b = `in`.readByte()
        }
        return str.toString()
    }

    @Throws(IOException::class)
    fun readValue(): String {
        val str = StringBuilder()
        var b = `in`.readByte()
        while (b != ','.code.toByte() && b != '.'.code.toByte()) {
            str.append(Char(b.toUShort()))
            b = `in`.readByte()
        }
        return str.toString()
    }

    @Throws(IOException::class)
    fun readString(): String {
        val str = StringBuilder()
        var b = `in`.readByte()
        while (b.toInt() != 0) {
            str.append(Char(b.toUShort()))
            b = `in`.readByte()
        }
        return str.toString()
    }

    @Throws(IOException::class)
    fun readUUID(): UUID {
        return UUID(`in`.readLong(), `in`.readLong())
    }

    @Throws(IOException::class)
    fun readInt(): Int {
        return `in`.readInt()
    }

    @Throws(IOException::class)
    fun readLong(): Long {
        return `in`.readLong()
    }

    @Throws(IOException::class)
    fun readByte(): Byte {
        return `in`.readByte()
    }

    @Throws(IOException::class)
    fun readBoolean(): Boolean {
        return `in`.readBoolean()
    }

    @Throws(IOException::class)
    fun readJson(): String {
        val json = StringBuilder()
        var b = `in`.readByte()
        if (b != '{'.code.toByte()) {
            return ""
        }
        json.append(Char(b.toUShort()))
        var num = 1
        while (num != 0) {
            b = `in`.readByte()
            if (b == '{'.code.toByte()) {
                num++
            } else if (b == '}'.code.toByte()) {
                num--
            }
            json.append(Char(b.toUShort()))
        }

        // Read the 0 appended to the end
        if (`in`.readByte().toInt() != 0) {
            throw RuntimeException("Expected null byte termination")
        }
        return json.toString()
    }

    val isClosed: Boolean
        get() = inSocket.isClosed

    @Throws(IOException::class)
    fun close() {
        out.close()
        `in`.close()
        inSocket.close()
    }
}